core/model/src/main/java/org/eclipse/rdf4j/model/util/Literals.java

```diff
@@ -57,6 +58,22 @@ public static String getLabel(Optional v, String fallback) {
 		return v != null ? getLabel((Value) v.orElseGet(null), fallback) : fallback;
 	}
 
+	/**
+	 * Retrieves the {@link XSD.Datatype} value for the supplied Literal, if it has one.
+	 *
+	 * @param a Literal
+	 * @return an Optional {@link XSD.Datatype} enum, if one is available. Note that the absence of this enum does
+	 *         <i>not</i> indicate that the literal has no datatype, merely that it has no cached enum representation of
+	 *         that datatype.
+	 * @since 3.5.0
+	 */
+	public static Optional<XSD.Datatype> getXsdDatatype(Literal l) {
+		if (l instanceof SimpleLiteral) {
+			return ((SimpleLiteral) l).getXsdDatatype();
+		}
+		return Optional.empty();
+	}
```

Comment:
Can we move this back into the Literal interface. There is no need to introduce another instanceof call for every Literal again!

Reply:
I'll run some benchmarks now to see the effects, but I understand that's it's hard to move back since it makes the Literal class dependent on the XSD class.

Reply:
This branch: 

```
Benchmark                                       Mode  Cnt    Score   Error  Units
ValueComparatorBenchmark.sortManyDeepEquals     avgt   10  330.260 ± 2.762  ms/op
ValueComparatorBenchmark.sortManyPointerEquals  avgt   10   19.662 ± 0.335  ms/op
ValueComparatorBenchmark.sortObjects            avgt   10   26.762 ± 0.800  ms/op
ValueComparatorBenchmark.sortPredicates         avgt   10    0.049 ± 0.016  ms/op
ValueComparatorBenchmark.sortSubjects           avgt   10   24.441 ± 0.448  ms/op
```

Develop: 
```
Currently running benchmarks
```

Reply:
No, we can't, because it uses a class from the model module, which introduces a cyclic dependency.

Reply:
Unless we can interface/abstract that out as well, somehow.

Reply:
Actually might not be a bad idea. It's late and I'm in the middle of a whisky tasting but I'll have a look tomorrow unless you best me to it.

Reply:
Enjoy your whisky tasting! No need to worry about this. I checked the performance, and using instanceof against a class seems to have no impact compared to how slow we were seeing instanceof be against an interface.